### ISSUE TITLE
Add a fold operation

### ISSUE THREAD

**Issue by tommymcm** [arcana-lab/memoir#77]
The fold will take a collection, an initial value, a function and an arbitrary number of closed arguments.

`memoir_fold(collection, initial_value, function, closure...)`

The instruction has the following semantics (in C++):

For sequences:
```c++
seq = ...
memoir_fold(init, seq, foo, closed...)
```

executes as:
```c++
auto accum = init;
for (size_t i = 0; i < memoir_size(seq); ++i) {
    auto val = memoir_read(seq, i);
    accum = foo(accum, i, val, closed...);
}
```

For associative collections:
```c++
assoc = ...
memoir_fold(init, assoc, foo, closed...)
```

executes as:
```c++
auto accum = init;
for (const auto [key, val] : assoc) {
    accum = foo(accum, key, val, closed...);
}
```

NOTE: At the moment, all closure arguments are immutable. To mutate a closed variable, it must be passed as a pointer, which will be loaded and stored in the function.

NOTE: the function used by a fold instruction _must_ be statically known, and _must_ be a top-level function in the module.

**Comment by tommymcm** [comment]
The core functionality of this is implemented and tested as of ce83a385266d198a2b2bbef8a6b478db9e8d5fa9.

Remaining work:
- [ ] Extend the type inference pass to inject type information for the fold function automatically.

**Comment by tommymcm** [comment]
The previous TODO has been completed as of c3c958eea75b94f1ea93e92a395a4db9ee0e0feb.
